Bug: multiple assignment from implicit array causes panic

The following works:

```ruby
» a, b, c = [1, 2, 3]
#» 3
» a
#» 1
» b
#» 2
» c
#» 3
```

The following causes panic:

```ruby
» a, b, c = 1, 2, 3
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x40f1455]

goroutine 1 [running]:
github.com/goby-lang/goby/compiler/ast.(*BaseNode).MarkAsExp(0x0)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/compiler/ast/ast.go:36 +0x5
github.com/goby-lang/goby/compiler/parser.(*Parser).parseStatement(0xc420138240, 0x486f768, 0x1)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/compiler/parser/statement_parsing.go:34 +0xbd
github.com/goby-lang/goby/compiler/parser.(*Parser).ParseProgram(0xc420138240, 0x11, 0xc42011e040)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/compiler/parser/parser.go:193 +0xe8
github.com/goby-lang/goby/igb.StartIgb(0x456d9c0, 0x5)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/igb/repl.go:160 +0x4d1
main.main()
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/goby.go:29 +0x954
```

FYI: works in Ruby:

```ruby
#ruby
i, j, k = 1, 2, 3
=> [1, 2, 3]
```

--- a/compiler/parser/expression_parsing_test.go
+++ b/compiler/parser/expression_parsing_test.go
@@ -690,3 +690,19 @@ func TestArithmeticExpressionFail(t *testing.T) {
 		}
 	}
 }
+
+// If parser doesn't crash then we covered panic successfully
+
+func TestMultipleAssignError(t *testing.T) {
+	input := `
+	a = 4, 5`
+
+	l := lexer.New(input)
+	p := New(l)
+	_, err := p.ParseProgram()
+
+	if err.Message != "unexpected 5 Line: 1" {
+		t.Fatal(err.Message)
+	}
+
+}

--- a/compiler/parser/expression_parsing_test.go
+++ b/compiler/parser/expression_parsing_test.go
@@ -691,8 +691,7 @@ func TestArithmeticExpressionFail(t *testing.T) {
 	}
 }
 
-// If parser doesn't crash then we covered panic successfully
-
+// After covered panic still return error to repl
 func TestMultipleAssignError(t *testing.T) {
 	input := `
 	a = 4, 5`
@@ -702,6 +701,21 @@ func TestMultipleAssignError(t *testing.T) {
 	_, err := p.ParseProgram()
 
 	if err.Message != "unexpected 5 Line: 1" {
+		t.Fatal(err.Message)
+	}
+
+}
+
+// Transfer the unexpected panic into error
+func TestUnexpectedPanicError(t *testing.T) {
+	input := `
+	3()`
+
+	l := lexer.New(input)
+	p := New(l)
+	_, err := p.ParseProgram()
+
+	if err.Message != "Some panic happen token: (. Line: 1" {
 		t.Fatal(err.Message)
 	}
 

--- a/compiler/parser/parser.go
+++ b/compiler/parser/parser.go
@@ -111,11 +111,15 @@ func New(l *lexer.Lexer) *Parser {
 }
 
 // ParseProgram update program statements and return program
-func (p *Parser) ParseProgram() (*ast.Program, *errors.Error) {
+func (p *Parser) ParseProgram() (program *ast.Program,err *errors.Error) {
 
-	defer func() {
+	defer func(){
 		if recover() != nil {
-			fmt.Println("Parser recovered from panic")
+			err = p.error
+			if err == nil {
+				msg := fmt.Sprintf("Some panic happen token: %s. Line: %d", p.curToken.Literal, p.curToken.Line)
+				err = errors.InitError(msg, errors.MethodDefinitionError)
+			}
 		}
 	}()
 
@@ -123,7 +127,7 @@ func (p *Parser) ParseProgram() (*ast.Program, *errors.Error) {
 	// Read two tokens, so curToken and peekToken are both set.
 	p.nextToken()
 	p.nextToken()
-	program := &ast.Program{}
+	program = &ast.Program{}
 	program.Statements = []ast.Statement{}
 
 	for !p.curTokenIs(token.EOF) {

--- a/compiler/parser/parser.go
+++ b/compiler/parser/parser.go
@@ -111,9 +111,9 @@ func New(l *lexer.Lexer) *Parser {
 }
 
 // ParseProgram update program statements and return program
-func (p *Parser) ParseProgram() (program *ast.Program,err *errors.Error) {
+func (p *Parser) ParseProgram() (program *ast.Program, err *errors.Error) {
 
-	defer func(){
+	defer func() {
 		if recover() != nil {
 			err = p.error
 			if err == nil {

--- a/compiler/parser/parser.go
+++ b/compiler/parser/parser.go
@@ -118,7 +118,7 @@ func (p *Parser) ParseProgram() (program *ast.Program, err *errors.Error) {
 			err = p.error
 			if err == nil {
 				msg := fmt.Sprintf("Some panic happen token: %s. Line: %d", p.curToken.Literal, p.curToken.Line)
-				err = errors.InitError(msg, errors.MethodDefinitionError)
+				err = errors.InitError(msg, errors.SyntaxError)
 			}
 		}
 	}()

